interface/requests/wireheader.go

```diff
@@ -6,68 +6,165 @@ package requests
-type WireHeader struct {
+type contentType uint8
+type versionMajorType uint8
+type versionMinorType uint8
+type flagsType uint16
+type sessionType uint64
+type acceptType uint8
+
+const (
+	versionMajorOne  versionMajorType = 0x01
+	versionMinorZero versionMinorType = 0x00
+
+	flagsZero flagsType = 0x0000
+)
+
+func (f flagsType) isValid() bool {
+	return f == flagsZero
+}
+
+func (a acceptType) isValid() bool {
+	return a == acceptType(0x00) //nolint:gomnd // Only to be set on sent headers
+}
+
+const (
+	contentTypeProtobuf contentType = 0x00
+)
+
+func (c contentType) isValid() bool {
+	return c == contentTypeProtobuf
+}
+
+// magic number to indicate we have a wire header
+const magicNumber uint32 = 0x5EC0A710
+
+// wireHeader represents a request header
+type wireHeader struct {
 	magicNumber  uint32
 	hdrSize      uint16
-	versionMajor uint8
-	versionMinor uint8
-	Flags        uint16
-	Provider     uint8
-	Session      uint64
-	ContentType  uint8
-	AcceptType   uint8
-	AuthType     uint8
-	BodyLen      uint32
-	AuthLen      uint16
-	OpCode       uint32
-	Status       uint16
-	Reserved1    uint8
-	Reserved2    uint8
+	versionMajor versionMajorType
+	versionMinor versionMinorType
+	flags        flagsType
+	provider     ProviderID
+	session      sessionType
+	contentType  contentType
+	acceptType   acceptType
+	authType     auth.AuthenticationType
+	bodyLen      uint32
+	authLen      uint16
+	opCode       OpCode
+	Status       StatusCode
+	reserved1    uint8
+	reserved2    uint8
 }
 
-func (r *WireHeader) pack(buf *bytes.Buffer) error {
+func (r *wireHeader) pack(buf *bytes.Buffer) error {
+	// panic rather than error as this is internal api and this shouldn't happen
+	if buf == nil {
+		panic("buffer pointer is nil")
+	}
+
+	// Check values are correct before packing
+
 	r.magicNumber = magicNumber
 	r.hdrSize = requestHeaderSize
+
+	if err := r.checkForRequest(); err != nil {
+		return err
+	}
+
 	err := binary.Write(buf, binary.LittleEndian, r)
 	return err
 }
 
+func (r *wireHeader) checkForRequest() error {
+	if !isSupportedWireHeaderVersion(r.versionMajor, r.versionMinor) {
+		return fmt.Errorf("invalid version %v.%v", r.versionMajor, r.versionMinor)
+	}
+	if !r.flags.isValid() {
+		return fmt.Errorf("invalid flags %v", r.flags)
+	}
+	if !r.contentType.isValid() {
+		return fmt.Errorf("invalid content type %v", r.contentType)
+	}
+	if !r.acceptType.isValid() {
+		return fmt.Errorf("invalid accept type %v", r.acceptType)
+	}
+	if !r.authType.IsValid() {
+		return fmt.Errorf("invaliid auth type %v", r.authType)
+	}
+	return nil
+}
+
+func isSupportedWireHeaderVersion(maj versionMajorType, min versionMinorType) bool {
+	return maj == versionMajorOne && min == versionMinorZero
+}
+
 const (
 	buffBytes8Bit  int = 1
 	buffBytes16Bit int = 2
 	buffBytes32Bit int = 4
 	buffBytes64Bit int = 8
 )
 
-func (r *WireHeader) parse(buf *bytes.Buffer) error {
+func parseWireHeaderFromBuf(buf *bytes.Buffer) (*wireHeader, error) {
+	r := &wireHeader{}
```

Comment:
How does the memory management work here? To me it looks like you are creating a temporary empty `wireHeader` on the stack and assign `r` to its address. Then fill it with the values parsed and return `r`. But when leaving the function, isn't `r` essentially going to be a dangling pointer?
And similarly in line 82 of `response.go`: `r := &Response{}`

Reply:
this is worth a read: https://www.ardanlabs.com/blog/2017/05/language-mechanics-on-escape-analysis.html

TL;DR - the compiler detects if a pointer will escape the current function - if it will, it allocate the data structure on the heap, otherwise it is free to choose whether to allocate on stack or heap. (likely stack but no guarantees